src/DynamicRupture/Output/Base.cpp

```diff
@@ -214,77 +216,329 @@ void Base::updateElementwiseOutput() {
+  }
+}
+
+void Base::computeSlipAndRate(std::array<real, 6>& rotatedTraction,
+                              std::array<real, 6>& rotatedLocalStress) {
+  real sFactorPlus = local.waveSpeedsPlus->sWaveVelocity * local.waveSpeedsPlus->density;
+  real sFactorMinus = local.waveSpeedsMinus->sWaveVelocity * local.waveSpeedsMinus->density;
 
-      // Note: cos1**2 can be greater than 1.0 because of rounding errors -> min
-      double sin1 = std::sqrt(1.0 - std::min(1.0, cos1 * cos1));
-      sin1 = (scalarProd > 0) ? sin1 : -sin1;
+  local.srS =
+      -(1.0 / sFactorPlus + 1.0 / sFactorMinus) * (rotatedTraction[3] - rotatedLocalStress[3]);
+  local.srD =
+      -(1.0 / sFactorPlus + 1.0 / sFactorMinus) * (rotatedTraction[5] - rotatedLocalStress[5]);
+}
+
+void Base::computeSlipAndRate(const double* tangent1,
+                              const double* tangent2,
+                              const double* strike,
+                              const double* dip) {
+  local.srS = static_cast<real>(0.0);
+  local.srD = static_cast<real>(0.0);
 
-      auto* slip1 = reinterpret_cast<DrPaddedArrayT>(layer->var(dynRup->slip1));
-      auto* slip2 = reinterpret_cast<DrPaddedArrayT>(layer->var(dynRup->slip2));
+  for (size_t i = 0; i < 3; ++i) {
+    real factorMinus = (local.faceAlignedValuesMinus[7] * tangent1[i] +
+                        local.faceAlignedValuesMinus[8] * tangent2[i]);
 
-      slipVectors(DirectionID::Strike, level, i) =
-          cos1 * slip1[ltsId][nearestGaussPoint] - sin1 * slip2[ltsId][nearestGaussPoint];
+    real factorPlus = (local.faceAlignedValuesPlus[7] * tangent1[i] +
+                       local.faceAlignedValuesPlus[8] * tangent2[i]);
 
-      slipVectors(DirectionID::Dip, level, i) =
-          sin1 * slip1[ltsId][nearestGaussPoint] + cos1 * slip2[ltsId][nearestGaussPoint];
+    local.srS += factorMinus * strike[i] - factorPlus * strike[i];
+    local.srD += factorMinus * dip[i] - factorPlus * dip[i];
+  }
+}
+
+real Base::computeRuptureVelocity() {
+  auto* ruptureTime = (local.layer->var(drDescr->ruptureTime))[local.ltsId];
+  real ruptureVelocity = 0.0;
+
+  bool computeRuptureVelocity{true};
+  for (size_t point = 0; point < tensor::QInterpolated::Shape[0]; ++point) {
+    if (ruptureTime[point] == 0.0) {
+      computeRuptureVelocity = false;
     }
   }
 
-  if (type == OutputType::AtPickpoint) {
-    outputData.cachedTime[outputData.currentCacheLevel] = time;
-    outputData.currentCacheLevel += 1;
+  if (computeRuptureVelocity) {
+
+    int nPoly = CONVERGENCE_ORDER - 1;
+    auto nearestGpIndex = local.nearestGpIndex;
+    int i1 = int((nearestGpIndex - 1) / (nPoly + 2)) + 1;
+    int j1 = (nearestGpIndex - 1) % (nPoly + 2) + 1;
+    if (i1 == 1) {
+      i1 = i1 + 1;
+    } else if (i1 == (nPoly + 2)) {
+      i1 = i1 - 1;
+    }
+
+    if (j1 == 1) {
+      j1 = j1 + 1;
+    } else if (j1 == (nPoly + 2)) {
+      j1 = j1 - 1;
+    }
+    nearestGpIndex = (i1 - 1) * (nPoly + 2) + j1;
```

Comment:
This does only work for the `stroud` Quadrature rule. For other quad rules, we do need to find new rules to find the closest interior gauss point.

Reply:
Hi, have you already implemented it in Fortran?

Reply:
No, it does not exist in Fortran.

Reply:
I've added a remark about this below....